sdk/python/kfp/dsl/_pipeline_volume.py

```diff
@@ -0,0 +1,407 @@
+
+
+class PipelineVolume(object):
+    """Representing a volume that is passed between pipeline operators.
+    That volume could be either new (created according to the user's needs from
+    scratch or from a VolumeSnapshot) or refering to an existing pvc.
+
+    A PipelineVolume object can be used as an extention of the pipeline
+    function's filesystem. It may then be passed between pipeline functions,
+    exposing dependencies.
+    """
+    def __init__(self, name: str = None, pvc: str = None,
+                 size: str = None, storage_class: str = None,
+                 data_source=None, mode: List[str] = VOLUME_MODE_RWM,
+                 annotations: Dict = None,
+                 k8s_resource: k8s_client.V1PersistentVolumeClaim = None,
+                 create: bool = True, is_exit_handler: bool = False):
+        """Create a new instance of PipelineVolume.
+        Args:
+            name: A desired name to refer to the PipelineVolume, and to use as
+                base when naming the created resource. This might change when
+                the pipeline gets compiled, in case the name is already added.
+            pvc: The name of an existing PVC
+            size: The size of the PVC which will be created
+            storage_class: The storage class to use for the dynamically created
+                PVC (requires size or data_source)
+            data_source: The name of an existing VolumeSnapshot or a
+                PipelineVolumeSnapshot object
+            mode: The access mode for the mounted PVC
+            annotations: Annotations to be patched on creating PVC
+            k8s_resource: An already configured k8s_resource
+            create: If a k8s_resource (PVC) is provided, this orders the
+                pipeline to create that resource.
+            is_exit_handler: Whether it is used as an exit handler.
+        Raises:
+        AttributeError: if pvc is set with size or data_source
+                        if storage_class is set along with pvc
+                        if storage_class is set without size or data_source
+                        if annotations are set without size or data_source
+                        if nothing of size, data_source and pvc is set
+        ValueError: if not inside a pipeline
+                    if k8s_resource without metadata.name or generateName is
+                        provided
+                    if size is an invalid memory string (when not a
+                        PipelineParam)
+                    if the name is an invalid string
+        """
+        if not _pipeline.Pipeline.get_default_pipeline():
+            raise ValueError("Default pipeline not defined.")
+
+        # Initialize
+        self.name = None
+        self.pvc = None
+        self.size = None
+        self.storage_class = None
+        self.data_source = None
+        self.data_source_name = None
+        self.mode = None
+        self.deps = set()
+        self.k8s_resource = None
```

Comment:
+1 on having a K8ResourceOp

Reply:
@hongye-sun 
> I really think it should have a base class named K8sResourceOp in parallel with ContainerOp to create general k8s resource. Volume and snapshot are two specialized ones that are deriving from it. A lot of code here can be reused.

@vicaire
>  +1 on having a K8ResourceOp
 
Hello @hongye-sun, @vicaire, thanks for the review!

This is a great suggestion. I am amending the design doc, to include your comment and a suggestion on how we can do it. We have tested it with an initial implementation of the code and it works beautifully. Essentially, it gives the user a way to create a custom K8s resource, from a resource spec, and have attributes of the K8s resource it creates be `PipelineParam`s for the rest of the pipeline.

Let's continue the discussion in the design doc, I will submit an extension shortly.

Reply:
@vkoukis 

The new design looks awesome and I really like it. It aligns with our vision of DSL and provides an elegant way to orchestrate arbitrary k8s resources. Let's continue our discussion on the design doc.

Reply:
btw, it's worth to take a look of https://github.com/kubeflow/pipelines/pull/879. In that PR, @eterna2 followed the same pattern to make ContainerOp and Container (which inherits from V1Container) as 2 different things, so that:
1. We are able to expose the full k8s container spec from ContainerOp
2. SideCar can inherit from Container
3. (in the future) We can create a TFJob which contains a Container. Then, all the convenient APIs that we support in Container like param parsing, volume mount, etc can be just reused.

That PR also implements a more generic way to extract PipelineParams from any object as long as it follows the convention from swagger generated code. This effort should make ResourceOp implementation a lot easier.

Reply:
@hongye-sun 
> The new design looks awesome and I really like it. It aligns with our vision of DSL and provides an elegant way to orchestrate arbitrary k8s resources. Let's continue our discussion on the design doc.
I think our visions for the DSL are aligned, it's just a matter of finding the most expressive/succinct way of describing our end goal. Your comments are very helpful in this regard. Looking forward to your comments on the design doc.